crypto/finite-field.ts

```diff
@@ -266,89 +266,112 @@ function createField(
       ? (x: bigint) => fastInverse(x, p, n, kmax, twoToMinusKmax)
       : (x: bigint) => inverse(x, p);
 
+  /**
+   * Ensures the given bigint is in its canonical form within the field.
+   * This function is added to address issues raised in a security audit,
+   * where certain operations assumed inputs were already in canonical form.
+   *
+   * It helps prevent incorrect results and infinite loops that could occur
+   * when non-canonical representations are used in operations like {@link isSquare()},
+   * {@link sqrt()}, and others. By using this function, we ensure consistent behavior
+   * across all field operations regardless of the input's initial representation.
+   */
+  function canonicalize(x: bigint) {
+    return x >= 0n && x < p ? x : mod(x, p);
+  }
+
   return {
     modulus: p,
     sizeInBits,
     t: oddFactor,
     M: twoadicity,
     twoadicRoot,
     mod(x: bigint) {
-      return mod(x, p);
+      return canonicalize(x);
     },
     add(x: bigint, y: bigint) {
-      return mod(x + y, p);
+      return canonicalize(x + y);
     },
     not(x: bigint, bits: number) {
-      return mod(2n ** BigInt(bits) - (x + 1n), p);
+      return canonicalize(2n ** BigInt(bits) - (x + 1n));
     },
     negate(x: bigint) {
-      return x === 0n ? 0n : p - x;
+      return x === 0n ? 0n : canonicalize(p - x);
     },
     sub(x: bigint, y: bigint) {
-      return mod(x - y, p);
+      return canonicalize(x - y);
     },
     mul(x: bigint, y: bigint) {
-      return mod(x * y, p);
+      return canonicalize(x * y);
     },
     inverse: exportedInverse,
     div(x: bigint, y: bigint) {
       let yinv = exportedInverse(y);
       if (yinv === undefined) return;
-      return mod(x * yinv, p);
+      return canonicalize(x * yinv);
     },
     square(x: bigint) {
-      return mod(x * x, p);
+      return canonicalize(x * x);
     },
     isSquare(x: bigint) {
-      return isSquare(x, p);
+      return isSquare(canonicalize(x), p);
     },
     sqrt(x: bigint) {
-      return sqrt(x, p, oddFactor, twoadicRoot, twoadicity);
+      const result = sqrt(
+        canonicalize(x),
+        p,
+        oddFactor,
+        twoadicRoot,
+        twoadicity
+      );
+      return result !== undefined ? canonicalize(result) : undefined;
     },
     power(x: bigint, n: bigint) {
-      return power(x, n, p);
+      return canonicalize(power(x, n, p));
     },
     dot(x: bigint[], y: bigint[]) {
       let z = 0n;
       let n = x.length;
       for (let i = 0; i < n; i++) {
-        z += x[i] * y[i];
+        z += canonicalize(x[i]) * canonicalize(y[i]);
       }
-      return mod(z, p);
+      return canonicalize(z);
     },
     equal(x: bigint, y: bigint) {
-      return mod(x - y, p) === 0n;
+      return canonicalize(x) === canonicalize(y);
```

Comment:
at first glance I thought this is less efficient than the original, but since inputs are typically canonical, it should actually be more efficient since we never have to `mod()`!

Reply:
Removed the `canonical` function and just replaced with inline checks to mod here! https://github.com/o1-labs/o1js-bindings/pull/287/commits/145a9286ab198a2156967bb8bdca6a45273827fa